zio/bzngio/index.go

```diff
@@ -0,0 +1,168 @@
+package bzngio
+
+import (
+	"errors"
+	"io"
+	"os"
+
+	"github.com/brimsec/zq/pkg/nano"
+	"github.com/brimsec/zq/zbuf"
+	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zng/resolver"
+)
+
+type TimeIndex struct {
+	index      []mark
+	indexReady bool
+}
+
+type mark struct {
+	Ts     nano.Ts
+	Offset int64
+}
+
+// NewIndex creates a new Index object, which is the container that holds
+// the in-memory index for a (b)zng file.  The first call to NewReader()
+// will return a reader that scans the entire file, building a time-based
+// index in the process, subsequent readers can use this index to read
+// only the relevant zng streams from the underlying file.
+func NewTimeIndex() TimeIndex {
+	return TimeIndex{}
+}
+
+// Create a new reader for the given zng file.  Only records with timestamps
+// that fall within the time range indicated by span will be emitted by
+// the returned Reader object.
+func (ti *TimeIndex) NewReader(f *os.File, zctx *resolver.Context, span nano.Span) (zbuf.ReadCloser, error) {
+	if ti.indexReady {
+		return newRangeReader(f, zctx, ti.index, span)
+	}
+
+	return &indexReader{
+		Reader: *NewReader(f, zctx),
+		Closer: f,
+		start:  span.Ts,
+		end:    span.End(),
+		parent: ti,
+	}, nil
+}
+
+// indexReader is a zbuf.Reader that also builds an index as it reads.
+type indexReader struct {
+	Reader
+	io.Closer
+	start   nano.Ts
+	end     nano.Ts
```

Comment:
Nit: Replace these with `span nano.Span`.

Reply:
This was deliberate since we reference end over and over and the only way to get the end time from a span is to compute it.  Maybe I'm optimizing prematurely or maybe the go compiler is clever enough to figure out that there's no aliasing and so that expression is constant and not recompute it every time, but I opted for the obviously-not-unnecessarily-inefficient route.

Reply:
Maybe do `end := i.span.End()` before entering the for loop inside Read, then?

Reply:
Sure, but Read() gets called repeatedly without the span ever changing.